Present information about when the token was last received

chriscray opened the pull request:
Here's a version of my proposed patch on master and with (hopefully) some clarification to address Chrissie Caulfield's comments on my previous patch (more explicit variable name, no longer overloading 'token_last_received')
Thanks!

chrissie-c commented:
Thanks. Honza is off till next week so we'll have to wait till then for him to review it too.

jfriesse commented:
Thanks for the patch. I like idea and implementation looks nice, but I see two issues:

1. Missing man page of new parameter. Could you please add it?
2. More problematic is static nature of `token_warning`. Used `token_timeout` reflects add/remove of nodes so it's kind of dynamic. Please consider rework `token_warning similarly.
3. If I understand patch correctly, it's NOP if `token_warning < token_timeout`. This should be documented.

Another possibility may be to change meaning of `token_warning` to be fraction (percentage) of `token_timeout`. That would also solve dynamic nature of token_timeout.

chriscray commented:
Thanks very much for the reviews.  I will take care of points 1 and 3.  I'm a little confused by point 2, however:
> More problematic is static nature of token_warning. Used token_timeout reflects add/remove of nodes so it's kind of dynamic. Please consider rework `token_warning similarly.

Are you worried about the case where, for example, token_timeout is set to 30000 and token_warning is set to 15000, and then a user adjusts token_timeout to 10000?

Thanks!

jfriesse commented:
@chriscray Yes, eventho I don't think opposite situation (token_timeout is enlarged) is much better because it may then warn user more often than needed. Also keep in mind that it's really not that much about user actively changing token timeout, but about adding/removing nodes into cluster. `token_timeout` is automatically increased/decreased, but `token_warning` (as implemented now) is not. That's why also suggested to use `token_warning` as a fraction of `token_timeout`

chriscray commented:
I pushed a new version that has token_warning as a percentage of the token timeout.  It also handles the case where a user dynamically turns off the token_warning.

Do you think I should print a warning/error when token_warning is less than token_hold_timeout?  In that case there will be a warning printed every token_warning interval since the token will never be received before token_hold_timeout.

Thanks
Chris

jfriesse commented:
@chriscray New patch looks really nice, so ACK by me.

Print warning when `token_hold_timeout` < `token_warning` seems like a good idea to me so if it is small amount of code then please add this feature.

chrissie-c commented:
I think that's a really good idea Honza. Something low like 75% sounds like it should catch problems and avoid false positives I hope, so sounds like a good starting point. As always with these things it really needs to be based on empirical data so we should probably keep an eye on it and adjust if necessary.

chriscray commented:
Great, thanks.  I pushed a new commit that:
1.  Changes the default to 75% (complete with man page change)
2.  Added a warning when the warning interval is less than token_retransmit_timeout.  The existing code only runs the token-related warnings (e.g., totem_config->token_timeout < MINIMUM_TIMEOUT) during runtime reconfiguration and not at startup (is this by design?), so I departed from this a little and added the token warning timeout to the totemsrp_initialize function in exec/totemsrp.c.
Thanks again!
Chris

jfriesse commented:
@chriscray Current code runs during startup, otherwise check
```
if (totem_config->token_warning > 100 || totem_config->token_warning < 0) {
```
couldn't work. It's just not possible to use logging that time, simply because during startup it's not fully configured yet, that's a reason why `snprintf` into `local_error_reason` is used. The code is not prepared for warnings, but I'm really not too keen to have a warning on two places, so please replace `log_printf` to `snprintf` stanza (just without `goto error_exit`). If there is problem in the config later, it will overwrite warning message, but that's not a problem really.

jfriesse commented:
@chriscray Eh, sorry, ignore my previous comment. snprintf into local_error_reason is not going to work. I have to find out something better.

jfriesse commented:
@chriscray Thank you for the update. I've squashed 3 patches together and merged as 3f7d2cf6aaea50dcd31ed4a16c13a1075fdb9ef8 .

Thank you for the PR.
